apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+            }
+        }
+    }
+
+    if (mBrushShape == CSVWidget::BrushShape_Circle)
+    {
+        for(int i = vertexCoords.first - r; i <= vertexCoords.first + r; ++i)
+        {
+            for(int j = vertexCoords.second - r; j <= vertexCoords.second + r; ++j)
+            {
+                cellId = CSMWorld::CellCoordinates::vertexGlobalToCellId(std::make_pair(i, j));
+                cellCoords = CSMWorld::CellCoordinates::fromId(cellId).first;
+                int distanceX = abs(i - vertexCoords.first);
+                int distanceY = abs(j - vertexCoords.second);
+                float distance = sqrt(pow(distanceX, 2)+pow(distanceY, 2));
+                int x = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(i);
+                int y = CSMWorld::CellCoordinates::vertexGlobalToInCellCoords(j);
+                float distancePerRadius = 1.0f * distance / r;
+                float smoothedByDistance = 0.0f;
+                if (mShapeEditTool == ShapeEditTool_Drag) smoothedByDistance = mTotalDiffY - mTotalDiffY * (3 * distancePerRadius * distancePerRadius - 2 * distancePerRadius * distancePerRadius * distancePerRadius);
+                if (mShapeEditTool == ShapeEditTool_PaintToRaise || mShapeEditTool == ShapeEditTool_PaintToLower) smoothedByDistance = (r + mShapeEditToolStrength) - (r + mShapeEditToolStrength) * (3 * distancePerRadius * distancePerRadius - 2 * distancePerRadius * distancePerRadius * distancePerRadius);
+                if (distance <= r)
+                {
+                    if (mShapeEditTool == ShapeEditTool_Drag) alterHeight(cellCoords, x, y, smoothedByDistance);
+                    if (mShapeEditTool == ShapeEditTool_PaintToRaise || mShapeEditTool == ShapeEditTool_PaintToLower)
+                    {
+                        alterHeight(cellCoords, x, y, smoothedByDistance);
+                        float smoothMultiplier = static_cast<float>(CSMPrefs::get()["3D Scene Editing"]["landedit-post-smoothstrength"].toDouble());
+                        if (CSMPrefs::get()["3D Scene Editing"]["landedit-post-smoothpainting"].isTrue()) smoothHeight(cellCoords, x, y, mShapeEditToolStrength * smoothMultiplier);
+                    }
+                    if (mShapeEditTool == ShapeEditTool_Smooth) smoothHeight(cellCoords, x, y, mShapeEditToolStrength);
+                    if (mShapeEditTool == ShapeEditTool_Flatten) flattenHeight(cellCoords, x, y, mShapeEditToolStrength, mTargetHeight);
+                }
+            }
+        }
+    }
+    if (mBrushShape == CSVWidget::BrushShape_Custom)
+    {
+        if(!mCustomBrushShape.empty())
+        {
+            for(auto const& value: mCustomBrushShape)
+            {
+                cellId = CSMWorld::CellCoordinates::vertexGlobalToCellId(std::make_pair(vertexCoords.first + value.first, vertexCoords.second + value.second));
+                cellCoords = CSMWorld::CellCoordinates::fromId(cellId).first;
```

Comment:
It's possible to avoid reusing `cellId` and `cellCoords` variables. First initialization is used only when `mBrushShape == CSVWidget::BrushShape_Point` is true. All other branches set a different values.

Reply:
True, worth a change? It does work how it should.

Reply:
Gave this one a proper look. It does improve the code a lot, so committed in 6f9f59d .